concurrency/TtsEngineSpeechBuffer.java

```diff
@@ -0,0 +1,198 @@
+import com.amazon.kcp.reader.utterance.SpeechBreakersUtteranceGetter;
+import com.amazon.kcp.reader.utterance.UtteranceGetter;
+import com.amazon.kindle.log.Log;
+import com.amazon.kindle.search.IKindleWordTokenIterator;
+import com.amazon.kindle.speech.breaker.ISpeechBreakerList;
+
+import java.text.BreakIterator;
+import java.util.*;
+
+/**
+ * This class encapsulates the buffer that the Tts engine will speak. It keeps
+ * track of what is being read, and tosses the sentence that was just uttered.
+ */
+public class TtsEngineSpeechBuffer {
+
+    private static final String TAG = TtsEngineSpeechBuffer.class.getName();
+
+    public static final String SPEAK_OPEN_TAG = "<speak>";
+    public static final String SPEAK_CLOSE_TAG = "</speak>";
+    public static final String CDATA_OPEN_TAG = "<![CDATA[";
+    public static final String CDATA_CLOSE_TAG = "]]>";
+    public static final String SSML_PARAGRAPH_OPEN = "<p>";
+    public static final String SSML_PARAGRAPH_CLOSE = "</p>";
+    public static final String SSML_BREAK = "<break/>";
+    public static final char NON_BREAKING_SPACE = '\u00A0';
+
+    public static final int NEXT_PAGE_INVALID = -1;
+
+    private BreakIterator wordBreakIterator;
+    private BreakIterator sentenceBreakIterator;
+
+    private TtsEngineDriver ttsEngineDriver;
+    private boolean pageFlipped;
+    private int pageStartPosition;
+    private int pageEndPosition;
+    private Map<Integer, MarkedUtterance> sentences;
+    private int progress = -1;
+
+    private String[] abbreviations;
+
+    private IKindleWordTokenIterator wordTokenIterator;
+    private ISpeechBreakerList speechBreakers = null;
+    private boolean speechBreakersAvailable;
+
+    private UtteranceGetter utteranceGetter;
+
+    public TtsEngineSpeechBuffer(TtsEngineDriver ttsEngineDriver) {
+        Log.debug(TAG, "<init>()");
+        this.ttsEngineDriver = ttsEngineDriver;
+        this.sentences = new TreeMap<Integer, MarkedUtterance>();
+    }
+
+    private void refreshSpeechBreakersAvailability() {
+        speechBreakersAvailable = (wordTokenIterator != null && speechBreakers != null && speechBreakers.getCount() > 0);
+    }
+
+    public synchronized void setWordTokenIterator(IKindleWordTokenIterator wordTokenIterator) {
+        if (this.wordTokenIterator != null) {
+            this.wordTokenIterator.close();
+        }
+        this.wordTokenIterator = wordTokenIterator;
+        refreshSpeechBreakersAvailability();
+    }
+
+    public void setSpeechBreakers(ISpeechBreakerList speechBreakers) {
+        this.speechBreakers = speechBreakers;
+        refreshSpeechBreakersAvailability();
+    }
+
+    public List<MarkedUtterance> readText(int startPosition, int endPosition, int currentUtteranceId) {
+        List<MarkedUtterance> utterancesParsed = new LinkedList<MarkedUtterance>();
+        if (sentences.size() == 0 || startPosition != pageStartPosition || endPosition != pageEndPosition) {
+            sentences.clear();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 This line may not be thread-safe: **81**. At this line the following field is not protected by synchronization: `sentences`. Your code typically protects this field with synchronization on object: `TtsEngineSpeechBuffer.this`.

**Fix**  
 Consider protecting the above field with synchronization on object: `TtsEngineSpeechBuffer.this`. For example, like at lines: **115**, **147**, **151**, **172**, **178**, **179**, **136**, **122**, **123**, and **129**.

**More info**  
 The following lines are where the above field is already protected by synchronization on the above object: **115**, **122**, **123**, **129**, **136**, **147**, **151**, **172**, **178**, and **179**.